compiler/src/org.graalvm.compiler.truffle.runtime/src/org/graalvm/compiler/truffle/runtime/debug/disassembler/MachineCode.java

```diff
@@ -0,0 +1,148 @@
+    private final Infopoint[] infopoints;
+
+    public static MachineCode read(long address, long headerLength, int size, Infopoint[] infopoints) {
+        return new MachineCode(address, headerLength, readCode(address, size), infopoints);
+    }
+
+    private MachineCode(long address, long headerLength, byte[] code, Infopoint[] infopoints) {
+        this.address = address;
+        this.headerLength = headerLength;
+        this.code = code;
+        this.infopoints = infopoints;
+    }
+
+    public int getLength() {
+        return code.length;
+    }
+
+    public String disassemble(DisassemblyFormatType disassemblyFormat) throws IOException {
+        return createDisassembler(disassemblyFormat).disassemble(new MachineCodeAccessor() {
+
+            @Override
+            public long getAddress() {
+                return address;
+            }
+
+            @Override
+            public long getHeaderLength() {
+                return headerLength;
+            }
+
+            @Override
+            public int getLength() {
+                return code.length;
+            }
+
+            @Override
+            public byte getByte(int n) {
+                return code[n];
+            }
+
+            @Override
+            public byte[] getBytes() {
+                return Arrays.copyOf(code, code.length);
+            }
+
+            @Override
+            public Stream<Infopoint> getInfopoints() {
+                return Arrays.stream(infopoints);
+            }
+
+            @Override
+            public String fileName(String extension) {
+                return String.format("truffle_compiled_code_%d_0x%x_%d%s", getPid(), address, System.nanoTime(), extension);
+            }
+
+        });
+    }
+
+    private Disassembler createDisassembler(DisassemblyFormatType disassemblyFormat) {
+        switch (disassemblyFormat) {
+            case HEX:
+                return new HexDisassembler();
+            case RAW:
+                return new RawDisassembler();
+            case ELF:
+                return new ElfDisassembler();
+            case OBJDUMP:
+                return new ObjdumpDisassembler();
+            default:
+                throw new UnsupportedOperationException();
+        }
+    }
+
+    private static long getPid() {
+        try {
+            final String info = java.lang.management.ManagementFactory.getRuntimeMXBean().getName();
+            return Long.parseLong(info.split("@")[0]);
+        } catch (Exception e) {
+            return 0;
+        }
+    }
+
+    private static byte[] readCode(long address, int size) {
+        final byte[] bytes = new byte[size];
+
+        for (int n = 0; n < bytes.length; n++) {
+            bytes[n] = UNSAFE.getByte(address + n);
+        }
+
+        return bytes;
+    }
+
+    private static final Unsafe UNSAFE = getUnsafe();
+
+    private static Unsafe getUnsafe() {
+        try {
+            Field f = Unsafe.class.getDeclaredField("theUnsafe");
```

Comment:
Better to use `GraalUnsafeAccess` than reflection.

Reply:
`GraalUnsafeAccess` uses reflection itself.

And I don't think I can access `GraalUnsafeAccess` from this class, can I? `GraalUnsafeAccess` is, like everything else, locked down to an extreme.